specification/dartLangSpec.tex

```diff
@@ -5193,52 +5195,70 @@ \subsubsection{Correct Member Overrides}
 \begin{itemize}
 \item
   $m$ and $m'$ are both methods, both getters, or both setters.
-  % We cannot have a setter and a method, say, because they cannot have
-  % the same name. However, we must _allow_ $m$ and $m'$ to be two setters.
+\item
+  If $m$ and $m'$ are both getters:
+  The return type of $m$ must be a subtype of the return type of $m'$.
 \item
   If $m$ and $m'$ are both methods or both setters:
   Let $F$ be the function type of $m$
   except that the parameter type is the built-in class \code{Object}
-  for each parameter of $m$ which has the modifier \COVARIANT.
+  for each parameter of $m$ which is covariant-by-declaration
+  (\ref{covariantParameters}).
   Let $F'$ be the function type of $m'$.
   $F$ must then be a subtype of $F'$.
 
   \commentary{%
-  The subtype requirement ensures that argument list shapes
-  that are admissible for an invocation of a method with signature $m'$
-  are also admissible for an invocation of a method with signature $m$.
-  For instance, $m'$ may accept 2 or 3 positional arguments,
-  and $m$ may accept 1, 2, 3, or 4 positional arguments, but not vice versa.
-  This is a built-in property of the function type subtype rules.
-  %
-  Note that a member signature differs from
-  an underlying syntactic declaration $D$ in a class $C$.
-  In particular,
-  a parameter in a member signature has the modifier \COVARIANT{}
-  if and only if the parameter is covariant-by-declaration
-  (\ref{covariantParameters}),
-  and that may be the case due to declarations in a supertype of $C$,
-  so that modifier need not be present in $D$.
-  %
-  There is an additional potential compile-time error associated with
-  a parameter which is covariant-by-declaration
-  (\ref{instanceMethods}).
-  But we cannot cover that here as a property of member overrides,
-  because it is concerned with declarations in all superinterfaces,
-  indirect as well as direct.%
+    The subtype requirement ensures that argument list shapes
+    that are admissible for an invocation of a method with signature $m'$
+    are also admissible for an invocation of a method with signature $m$.
+    For instance, $m'$ may accept 2 or 3 positional arguments,
+    and $m$ may accept 1, 2, 3, or 4 positional arguments, but not vice versa.
+    This is a built-in property of the function type subtype rules.
+  }
+
+  If $p'$ is a formal parameter in $m'$ which is covariant-by-declaration,
+  the corresponding parameter $p$ in $m$ must also be covariant-by-declaration
```

Comment:
Again, should covariant by generics be included for `m'`?
So if `m'` is covariant-by-generics, then `m` must accept any argument.

For `m`, you don't need to say that it must be covariant, because if it was covariant, you already changed its parameter type to `Object` above (line 5204), so it is guaranteed to satisfy "$p$ is a top type". You can just have:

> the type of the corresponding parameter $p$ in $m$ must be a top type (which it is if $p$ is is covariant-by-declaration or covariant-by-generics).

Reply:
We decided to make it a non-error when the parameter is covariant-by-class. It isn't specified when the implementations need to implicitly induce a forwarding method to ensure soundness, we just specify that actual soundness violations must be detected (so an implementation can generate type checks everywhere, or it can optimize for removing as many as possibly&mdash;and using a forwarding method here and there is one of the tools available to reduce the number of generated type checks).

> if `m'` is covariant-by-generics, then `m` must accept any argument.

No, but there must be a dynamic check on the actual parameter type. An implicitly induced forwarder can do that.

> you already changed its parameter type

No, not quite: The function type `$F$` was computed in this manner, but the member signatures `$m$` and `$m'$` are derived from the program with no modifications.

Reply:
When I say "accept any argument" I mean doing something semantically correct for it, rather than, well, *actually* accepting something of the wrong type. So bad phrasing, what I meant was "Must behave soundly for all arguments", which means properly throwing if an argument allowed by the covariant upcast is not valid for the actual implementation.

Reply:
I think we agree here, so I'll close this thread because we have enough threads already. ;)